### PR TITLE
[WIP] Big refactor

### PR DESCRIPTION
**Pull request by l2isbad**
let's get it started
Fix :#19
Fix :#21

### CODE REVIEW

File: godplugin/configs.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	MaxProcs   int             `yaml:"max_procs"`
+	Modules    map[string]bool `yaml:"modules"`
+}
+
+func (c *Config) Load(filename string) error {
+	file, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	err = yaml.NewDecoder(file).Decode(c)
```
Comment (Wing924) [inline godplugin/configs.go]:
```suggestion
	return yaml.NewDecoder(file).Decode(c)
```

File: godplugin/configs.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+)
+
+func NewConfig() *Config {
+	return &Config{
+		DefaultRun: true,
+		Enabled:    true,
+	}
+}
+
+type Config struct {
```
Comment (Wing924) [inline godplugin/configs.go]:
```suggestion
type GlobalConfig struct {
```

File: godplugin/configs.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	if run, ok := c.Modules[module]; ok {
+		return run
+	}
+	if explicit {
+		return false
+	}
+	return c.DefaultRun
+}
+
+type rawModConfig []yaml.MapSlice
```
Comment (Wing924) [inline godplugin/configs.go]:
```suggestion
type ModuleConfig struct {
  Global ModuleGlobalConfig `yaml:"global"`
  Jobs []JobConfig `yaml:"jobs"`
}

type ModuleGlobalConfig  struct {...}
type JobConfig struct {...}
```

File: godplugin/module_config.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+type moduleConfig struct {
+	Global rawConfig   `yaml:"global"`
+	Jobs   []rawConfig `yaml:"jobs"`
+}
+
+func (m *moduleConfig) load(filename string) error {
+	file, err := os.Open(filename)
+	defer file.Close()
```
Comment (Wing924) [inline godplugin/module_config.go]: move to below err check

File: godplugin/plugin.go [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+}
+
+var log = logger.New("plugin", "main")
+var validate = validator.New()
+
+func New() *Plugin {
+	return &Plugin{
+		modules:   make(modules.Registry),
+		loopQueue: make([]Job, 0),
+		checkCh:   make(chan Job),
```
Comment (Wing924) [inline godplugin/plugin.go]: `make(chan Job, 1)`
at least 1 item

File: godplugin/plugin.go [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+	Shutdown()
+}
+
+var log = logger.New("plugin", "main")
+var validate = validator.New()
+
+func New() *Plugin {
+	return &Plugin{
+		modules:   make(modules.Registry),
+		loopQueue: make([]Job, 0),
```
Comment (Wing924) [inline godplugin/plugin.go]: don't need it it.
loopQueue default init to empty slice

File: godplugin/plugin.go [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+	p.MainLoop()
+}
+
+func (p *Plugin) MainLoop() {
+	log.Info("start main loop")
+	var clock int
+	tk := ticker.New(time.Second)
+
+	for {
+		select {
```
Comment (Wing924) [inline godplugin/plugin.go]: we don't need select case here

File: godplugin/plugin.go [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+		}
+	}
+}
+
+func (p *Plugin) checkJobs() {
+	started := make(map[string]bool)
+
+	for job := range p.checkCh {
+		if started[job.FullName()] {
+			log.Warningf("skipping %s[%s]: already served by another job", job.ModuleName(), job.Name())
```
Comment (Wing924) [inline godplugin/plugin.go]: it's normal, not warning. info is ok

File: godplugin/plugin.go [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+	}
+}
+
+func (p *Plugin) createJobs() []Job {
+	var jobs []Job
+	for name, creator := range p.modules {
+		var modConfig moduleConfig
+
+		// FIXME:
+		err := modConfig.load(fmt.Sprintf("/opt/go.d/%s.conf", name))
```
Comment (Wing924) [inline godplugin/plugin.go]: hard code path

File: godplugin/plugin.go [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+
+		if err != nil {
+			log.Errorf("skipping %s: %s", name, err)
```
Comment (Wing924) [inline godplugin/plugin.go]: `"skipping %s: %v"`

File: godplugin/plugin.go [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+			continue
+		}
+
+		if len(modConfig.Jobs) == 0 {
+			log.Errorf("skipping %s: config 'Jobs' section is empty or not exist", name)
+			continue
+		}
+
+		jobName := func(conf rawConfig) interface{} {
```
Comment (Wing924) [inline godplugin/plugin.go]: why create a func in loop?

File: godplugin/module_config.go (hunk 4 above)
Comment (l2isbad) [reply godplugin/module_config.go]: reason?

File: godplugin/plugin.go (hunk 9 above)
Comment (l2isbad) [reply godplugin/plugin.go]: it's under FIXME, i know it

File: godplugin/plugin.go (hunk 11 above)
Comment (l2isbad) [reply godplugin/plugin.go]: actually for no reason lol

File: godplugin/module_config.go (hunk 4 above)
Comment (Wing924) [reply godplugin/module_config.go]: if err is not nil, file may be nil

Comment (l2isbad) [reply godplugin/module_config.go]: `Close` can handle it, since we are not checking `Close` result it's OK to leave it here

**Comment by l2isbad** [comment]
This is huge.

**Pull request merged**